src/core/geometry/qgsgeometrycollection.h

```diff
@@ -334,7 +334,13 @@ class CORE_EXPORT QgsGeometryCollection: public QgsAbstractGeometry
      */
     inline static const QgsGeometryCollection *cast( const QgsAbstractGeometry *geom )
     {
-      if ( geom && QgsWkbTypes::isMultiType( geom->wkbType() ) )
+      if ( !geom )
+        return nullptr;
+
+      const Qgis::WkbType flatType = QgsWkbTypes::flatType( geom->wkbType() );
+      if ( QgsWkbTypes::isMultiType( geom->wkbType() )
```

Comment:
I would argue that isMultiType() should not return true for PolyhedralSurface and TIN if they can't be cast to GeometryCollection. At least that is what GDAL does

Reply:
While working on it, I was not sure on the correct approach. Since, we're trying to have an approach similar to GDAL. This should be changed. Thanks for the suggestion.

Reply:
Yeah, there's a TON of places where we check if a geometry isMultiType and then assume it can be safely static_cast as a collection. Let's not break that assumption here.

Reply:
In `WkbEntries`, one need to define the `multiType` of `polyhedraSurface` and `TIN`. What would be the appropriate multi type: a `QgsMultiPolygon`, a `QgsGeometryCollection`? 

SFCGAL uses a `Solid`, `MultiSolid` logic but we do not have this logic in QGIS.

Reply:
I'd say multipolygon

Reply:
`QgsPolyhedralSurface` and `QgsTriangulatedSurface` are now single types. I have also removed the logic to consider `TIN` as the multi type of `QgsTriangle`.